nemo_aligner/models/nlp/gpt/megatron_gpt_dpo_model.py

```diff
@@ -190,20 +221,62 @@ def get_reduced_masked_logps(self, logps, labels, average_log_probs=False):
         else:
             return (logps * loss_mask).sum(-1)
 
-    def loss_func(self, pi_logprobs, ref_logprobs, labels, average_log_probs=False):
+    def loss_func(self, pi_logprobs, ref_logprobs, labels, gt_rewards, average_log_probs=False):
         rewards = self.get_reduced_masked_logps(
             pi_logprobs - ref_logprobs, labels, average_log_probs=average_log_probs
         )
-        chosen_rewards, reject_rewards = self.split_output_tensor(self.ref_policy_kl_penalty * rewards)
-
-        loss = -torch.nn.functional.logsigmoid(chosen_rewards - reject_rewards).mean(0)
+        chosen_rewards, reject_rewards = self.split_output_tensor(rewards)
+        rewards_delta = chosen_rewards - reject_rewards
+
+        if self.preference_loss == "dpo":
+            loss = -torch.nn.functional.logsigmoid(self.ref_policy_kl_penalty * rewards_delta).mean(0)
+        elif self.preference_loss == "reward_bwd_dpo":
+            logbeta_hat_chosen = torch.nn.functional.logsigmoid(self.ref_policy_kl_penalty * rewards_delta)
+            logbeta_hat_rejected = torch.nn.functional.logsigmoid(-self.ref_policy_kl_penalty * rewards_delta)
+
+            chosen_gt_rewards, reject_gt_rewards = self.split_output_tensor(gt_rewards)
+            gt_rewards_delta = self.gt_reward_scale * (chosen_gt_rewards - reject_gt_rewards)
```

Comment:
How sensitive are the downstream results to this hyperparameter? Could we maybe derive it somehow rather than making it an unbounded parameter which has to be hand-tuned?

Reply:
It is quite stable. In my experiment, I just used it as 1. Depending on the ground-truth reward scales, you might want to tune it larger or smaller, but not much.

Reply:
But if you always used 1.0, then is there any reason really to even have this parameter?If 90% of the time it's just 1.0 anyway

Reply:
It is still useful. Imagine a case, if all the predicted rewards are in the range of 0-1  (or 0-0.1 more extremely), we want to scale up the rewards properly otherwise the policy network cannot learn. In fact, I do want to tune this hyperparameter in the upcoming ablation study.